## Arithmetic Operations

Python has the capability to perform calculations involving addition, subtraction, multiplication, and division. These basic arithmetic operations are some of the simplest types of commands that Python is able to process. We will discuss several of these operations in this section. 


## Addition and Subtraction

Like any good discussion of mathematical operations, our will begin with addition and subtraction. In the celll below, we provide an example of using Python to perform addition. 

In [1]:
3 + 14

17

We can also perform subtraction in Python. 

In [2]:
3 - 15

-12

## Multiplication and Division 

We will now escalate things by considering the operations of multiplcation and division. Multiplication is performed using the `*` operator. 

In [3]:
6 * 7

42

In the next three cells, we provide examples of division. 

In [3]:
5 / 2

2.5

In [5]:
8 / 2

4.0

In [6]:
7 / 3

2.3333333333333335

If we attempt to divide by zero, we get an error. 

In [13]:
42 / 0

ZeroDivisionError: division by zero

Errors are an ever-present part of any programmer's existence. Every error results in an error message that attempts to explain what caused the error. Note that the last line in the output above says: 

    ZeroDivisionError: division by zero
    
This is Python's way of telling us that we got an error because we attempted to commmit the mathematical sin of dividing by zero. Error messages are often cryptic and difficult to understand, but it is important that you attempt to understand what any error messages you encounter are trying to tell you. Being able to read error messages can be a big help when attempting to debug your code. 

## Exponentiation

Python uses two asterisks, `**` to indicate the operation of exponentiation. In the next cell, we will calculate 3 raised to the 5th power, or $3^5$.

In [23]:
3 ** 5

243

Many programming languages and computer algebra systems use the "carat" symbol `^` for exponentiation. Python is somewhat unusual in its use of `**` for exponentiation. However, even if you are familiar with using `^` to indicate exponents, you will likely get used to Python's way of doing things quickly. 

Be careful, though. Python *does* use the carat operator, but to indicate a different operation called bitwise XOR (which we won't discuss here). This can be something of a pitfall because if you typed `3 ^ 5` in an attempt to calculate $3^5$, you would NOT get an error, but you would get the incorrect answer (try this on your own). If you didn't realize that your answer was wrong, that could cause some problems for you down the road. 

## The Modulus Operator

Python, like most programming languages, provides the ability to perform the **modulus operation**. This operation provides the remainder that would result from dividing one integer by another. In Python, this is performed by playing the `%` symbol between two integers. The number on the left is the dividend and the one on the right is the divisor. We provide several examples below. 

In [26]:
8 % 3

2

In [27]:
25 % 7

4

In [28]:
90 % 3

0

The modulus operator is often useful to programmers. We will explore some of its applications in a later section. 

Several programming languages use `%` to indicate the modulus operator, but it is also common to see `%%` or `mod`. 

## Floor Division

A mathematical operation related to the modulus is **floor division**. This operation can be performed on two integers, and will give the integer result of dividing one integer by the other. You can think of this as first performing regular division, and then rounding down to the next largest whole number. Floor division is performed using the symbols `//`. 

In [29]:
8 // 3

2

In [30]:
25 // 7

3

In [31]:
90 // 3

30

## Order of Operations

Python recognizes the standard order of operations when performing calculations. This is demonstrated in the cell below. 

In [8]:
4 + 5 * 6

34

Parentheses can be used to form more complex arithmetic expressions. 

In [33]:
5 * (3 + 8 / 2)

35.0

## Rounding Issues

The result obtained by dividing 7 by 3 is 2.3333.... Notice that when we performed this operation above, we got `2.3333333333333335`. It might seem a bit odd that Python decided to round the last digit of that number of a 5. You will encounter rounding odditities like this from time to time. In fact, the next example will likely seem even stranger.

In [14]:
0.1 + 0.2

0.30000000000000004

These unexpected results stem from the fact that computers represent numbers in a base 2 (binary) number system rather than the base 10 (decimal) system that modern humans are familiar with. While it might be disconcerting to encounter results like this, the good news is that the difference between the expected result and what we were give by Python is extremely small, and is incredibly unlikely to be any cause for concern. Later on, we will learn how to "hide" this sort of behavior by rounding the results of our calculations. 